haskell.nix

```
@@ -0,0 +1,110 @@
+          haskellPackages = mkOption {
+            type = types.anything;
+            description = ''Which Haskell package set to use'';
+            default = pkgs.haskellPackages;
+          };
+          name = mkOption {
+            type = types.str;
+            description = ''Name of the Cabal package'';
+          };
+          root = mkOption {
+            type = types.path;
+            description = ''Path to the Cabal project root'';
+            default = ./.;
+          };
+          overrides = mkOption {
+            type = types.anything;
+            description = ''Overrides for the Cabal project'';
+            default = self: super: { };
+          };
+          modifier = mkOption {
+            type = types.anything;
+            description = ''Modifier for the Cabal project'';
+            default = drv: drv;
+          };
+          baseBuildTools = mkOption {
+            type = types.anything;
+            description = ''Common build tools for Haskell develop; you don't need to set this.'';
+            default = hp: with hp; [
+              cabal-install
+              haskell-language-server
+              ghcid
+              hlint
+              cabal-fmt
+            ];
+          };
+          extraBuildTools = mkOption {
+            type = types.anything;
+            description = ''Extra tools to add to nix-shell'';
+            default = hp: [ ];
+          };
+        };
+      });
+  };
+  config = {
+    perSystem = { config, self', inputs', pkgs, ... }:
+      let
+        cfg = config.haskellProject;
+        inherit (pkgs.lib.lists) optionals;
+
+        # Specify GHC version here. To get the appropriate value, run:
+        #   nix-env -f "<nixpkgs>" -qaP -A haskell.compiler
+        hp = cfg.haskellPackages; # pkgs.haskellPackages; # Eg: pkgs.haskell.packages.ghc921;
+
+        buildTools = cfg.baseBuildTools hp ++ cfg.extraBuildTools hp;
+
+        project =
+          { returnShellEnv ? false
+          , withHoogle ? false
+          }:
+          hp.developPackage {
+            inherit returnShellEnv withHoogle;
+            inherit (cfg) root name overrides;
+            modifier = drv:
+              cfg.modifier (pkgs.haskell.lib.overrideCabal drv (oa: {
+                # All the Cabal-specific overrides go here.
+                # For examples on what is possible, see:
+                #   https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/lib/compose.nix
+                buildTools = (oa.buildTools or [ ]) ++ optionals returnShellEnv buildTools;
+              }));
+          };
+      in
+      {
+        # Used by `nix build ...`
+        packages = {
+          ${cfg.name} = project { };
+          default = mkDefault self'.packages.${cfg.name};
+        };
+        # Used by `nix run ...`
+        apps = {
+          ${cfg.name} = {
+            type = "app";
+            program = pkgs.lib.getExe self'.packages.default;
+          };
```

```suggestion
          ${cfg.name} = {
            program = self'.packages.default;
          };
```

`type` is already handled by the `flake-parts` core modules.

`program` also performs `getExe` using `types.coercedTo`.